### ISSUE TITLE
Casrel模型代码片段请教

### ISSUE THREAD

**Issue by Joy-2000** [taishan1994/pytorch_casrel_triple_extraction#4]
您好，Casrel模型中使用sigmoid计算subjects/objects preds时，能否向您请教一下这里分别加2次和4次方计算是为什么？
`subject_preds = (torch.sigmoid(self.linear1(seq_output)))**2`
`output = (torch.sigmoid(self.linear2(output)))**4`
```
class Casrel(nn.Module):
    def __init__(self, args, tag2id):
        super().__init__()
        self.bert = BertModel.from_pretrained(args.bert_dir)
        self.tag2id = tag2id
        self.linear1 = nn.Linear(768, 2)
        # 768*2
        self.condLayerNorm = LayerNorm(hidden_size=768, conditional_size=768*2)
        self.linear2 = nn.Linear(768, len(tag2id)*2)
        self.crierion = MyBCELoss()
    
def forward(self, 
          token_ids, 
          attention_masks, 
          token_type_ids,
          subject_labels=None,
          object_labels=None,
          subject_ids=None):
        # 预测subject
        bert_outputs = self.bert(
            input_ids=token_ids,
            attention_mask=attention_masks,
            token_type_ids=token_type_ids
        )  
        seq_output = bert_outputs[0]  # [btz, seq_len, hdsz]
        subject_preds = (torch.sigmoid(self.linear1(seq_output)))**2 # [btz, seq_len, 2]

        # 传入subject，预测object
        # 通过Conditional Layer Normalization将subject融入到object的预测中
        # 理论上应该用LayerNorm前的，但是这样只能返回各个block顶层输出，这里和keras实现不一致
        subject = self.extract_subject([seq_output, subject_ids])
        output = self.condLayerNorm([seq_output, subject])
        output = (torch.sigmoid(self.linear2(output)))**4
        object_preds = output.reshape(*output.shape[:2], len(self.tag2id), 2)
        # print(object_preds.shape, object_labels.shape)
        loss = self.crierion([subject_preds, object_preds], [subject_labels, object_labels, attention_masks])
        return loss
```

**Comment by taishan1994** [comment]
> 您好，Casrel模型中使用sigmoid计算subjects/objects preds时，能否向您请教一下这里分别加2次和4次方计算是为什么？ `subject_preds = (torch.sigmoid(self.linear1(seq_output)))**2` `output = (torch.sigmoid(self.linear2(output)))**4`
> 
> ```
> class Casrel(nn.Module):
>     def __init__(self, args, tag2id):
>         super().__init__()
>         self.bert = BertModel.from_pretrained(args.bert_dir)
>         self.tag2id = tag2id
>         self.linear1 = nn.Linear(768, 2)
>         # 768*2
>         self.condLayerNorm = LayerNorm(hidden_size=768, conditional_size=768*2)
>         self.linear2 = nn.Linear(768, len(tag2id)*2)
>         self.crierion = MyBCELoss()
>     
> def forward(self, 
>           token_ids, 
>           attention_masks, 
>           token_type_ids,
>           subject_labels=None,
>           object_labels=None,
>           subject_ids=None):
>         # 预测subject
>         bert_outputs = self.bert(
>             input_ids=token_ids,
>             attention_mask=attention_masks,
>             token_type_ids=token_type_ids
>         )  
>         seq_output = bert_outputs[0]  # [btz, seq_len, hdsz]
>         subject_preds = (torch.sigmoid(self.linear1(seq_output)))**2 # [btz, seq_len, 2]
> 
>         # 传入subject，预测object
>         # 通过Conditional Layer Normalization将subject融入到object的预测中
>         # 理论上应该用LayerNorm前的，但是这样只能返回各个block顶层输出，这里和keras实现不一致
>         subject = self.extract_subject([seq_output, subject_ids])
>         output = self.condLayerNorm([seq_output, subject])
>         output = (torch.sigmoid(self.linear2(output)))**4
>         object_preds = output.reshape(*output.shape[:2], len(self.tag2id), 2)
>         # print(object_preds.shape, object_labels.shape)
>         loss = self.crierion([subject_preds, object_preds], [subject_labels, object_labels, attention_masks])
>         return loss
> ```

苏剑林应该有个博客是讲这个的，具体是哪篇忘了。直观来讲假设输出为[0.9, 0.1]，**2之后变为[0.81, 0.01]，相当于变相抑制了错误类别。

**Comment by Joy-2000** [comment]
好的 非常感谢！
